src/virtual-machine.js

```diff
@@ -769,19 +770,51 @@ class VirtualMachine extends EventEmitter {
         targets = targets.filter(target => !!target);
 
         return this._loadExtensions(extensions.extensionIDs, extensions.extensionURLs).then(() => {
+            const deprecated_extensionStorage = {};
             for (const extension of extensions.extensionIDs) {
-                if (`ext_${extension}` in this.runtime) {
-                    if ((typeof this.runtime[`ext_${extension}`].deserialize === 'function') && 
-                        extensions.extensionData[extension]) {
-                        this.runtime[`ext_${extension}`].deserialize(extensions.extensionData[extension]);
-                    }
+                if (!extensions.extensionData[extension]) continue;
+                if (
+                    `ext_${extension}` in this.runtime &&
+                    typeof this.runtime[`ext_${extension}`].deserialize === 'function'
+                ) {
+                    this.runtime[`ext_${extension}`].deserialize(extensions.extensionData[extension]);
+                    continue;
                 }
+                deprecated_extensionStorage[extension] = extensions.extensionData[extension];
             }
+            if (deprecated_extensionStorage)
+                this.runtime.extensionStorage = ExtensionStorage(deprecated_extensionStorage);
+
             targets.forEach(target => {
                 this.runtime.addTarget(target);
                 (/** @type RenderedTarget */ target).updateAllDrawableProperties();
                 // Ensure unique sprite name
                 if (target.isSprite()) this.renameSprite(target.id, target.getName());
+
+                if (!("extensionData" in target)) return;
+
+                const deprecated_extensionStorage_pertarget = {};
+
+                for (const extension of extensions.extensionIDs) {
+                    if (!(extension in target.extensionData)) continue;
+
+                    if (
+                        `ext_${extension}` in this.runtime &&
+                        typeof this.runtime[`ext_${extension}`].deserializeForTarget === 'function'
+                    ) {
+                        this.runtime[`ext_${extension}`].deserializeForTarget(
+                            target.extensionData[extension],
+                            target,
+                        );
+                        continue;
+                    }
+                    deprecated_extensionStorage_pertarget[extension] = target.extensionData[extension];
+                }
+
+                if (deprecated_extensionStorage)
+                    target.extensionStorage = ExtensionStorage(deprecated_extensionStorage_pertarget);
```

Comment:
I genuinely forgot why I decided to make this partly snake case and partly camel case.